builtin/logical/aws/path_roles.go

```diff
@@ -58,94 +85,342 @@ func pathRoles() *framework.Path {
+		}
+		roleEntry.RoleArns = roleArnsRaw.([]string)
+	}
+
+	if policyArnsRaw, ok := d.GetOk("policy_arns"); ok {
+		if legacyRole != "" {
+			return logical.ErrorResponse("cannot supply deprecated role or policy parameters with policy_arns"), nil
+		}
+		roleEntry.PolicyArns = policyArnsRaw.([]string)
+	}
+
+	if policyDocumentRaw, ok := d.GetOk("policy_document"); ok {
+		if legacyRole != "" {
+			return logical.ErrorResponse("cannot supply deprecated role or policy parameters with policy_document"), nil
+		}
+		var compacted string
+		if policyDocumentRaw.(string) == "" {
+			compacted = ""
+		} else {
+			compacted, err = compactJSON(policyDocumentRaw.(string))
+			if err != nil {
+				return logical.ErrorResponse(fmt.Sprintf("cannot parse policy document: %q", policyDocumentRaw.(string))), nil
+			}
+		}
+		roleEntry.PolicyDocument = compacted
+	}
+
+	if legacyRole != "" {
+		roleEntry = upgradeLegacyPolicyEntry(legacyRole)
+		if roleEntry.InvalidData != "" {
+			return logical.ErrorResponse(fmt.Sprintf("unable to parse supplied data: %q", roleEntry.InvalidData)), nil
+		}
+		resp.AddWarning("Detected use of legacy role or policy paramemter. Please upgrade to use the new parameters.")
+	} else {
+		roleEntry.ProhibitFlexibleCredPath = false
+	}
+
+	if len(roleEntry.CredentialTypes) < 1 {
+		return logical.ErrorResponse("did not supply credential_type"), nil
+	}
 
-	if !bp && !ba {
-		return false, errors.New("either policy or arn must be provided")
+	if len(roleEntry.RoleArns) > 0 && !strutil.StrListContains(roleEntry.CredentialTypes, assumedRoleCred) {
+		return logical.ErrorResponse(fmt.Sprintf("cannot supply role_arns when credential_type isn't %s", assumedRoleCred)), nil
 	}
-	if bp && ba {
-		return false, errors.New("only one of policy or arn should be provided")
+	if len(roleEntry.PolicyArns) > 0 && !strutil.StrListContains(roleEntry.CredentialTypes, iamUserCred) {
+		return logical.ErrorResponse(fmt.Sprintf("cannot supply policy_arns when credential_type isn't %s", iamUserCred)), nil
 	}
-	return bp, nil
+
+	err = nonLockedSetAwsRole(ctx, req.Storage, roleName, roleEntry)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp, nil
 }
 
-func pathRolesWrite(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	var buf bytes.Buffer
+func nonLockedRoleRead(ctx context.Context, s logical.Storage, roleName string) (*awsRoleEntry, error) {
+	if roleName == "" {
+		return nil, fmt.Errorf("missing role name")
+	}
+	entry, err := s.Get(ctx, "role/"+roleName)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result awsRoleEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+	return &result, nil
+}
+
+func (b *backend) lockedRoleRead(ctx context.Context, s logical.Storage, roleName string) (*awsRoleEntry, error) {
+	b.roleMutex.RLock()
+	roleEntry, err := nonLockedRoleRead(ctx, s, roleName)
```

Comment:
Based on some of the other suggestions, the need for the nonLocked method may go away.  If so, can you just inline this logic.